lib/factor-rules.js

```diff
@@ -0,0 +1,373 @@
+
+// TODO: add restrictions (#c_0 divisible by 2 && #a_0 is perfect square)
+
+// e.g. 4x^2 - 9y^2 -> (2x)^2 - (3y)^2
+export const FACTOR_DIFFERENCE_OF_SQUARES_HELPER =
+  defineRuleString('#a #b^#c - #d #e^#f', '(#eval(nthRoot(#a)) #b^(#eval(#c/2)))^2 - (#eval(nthRoot(#d)) #e^(#eval(#f/2)))^2')
+
+// e.g. (2x)^2 - (3y)^2 -> (2x + 3y)(2x - 3y)
+export const FACTOR_DIFFERENCE_OF_SQUARES =
+  defineRuleString('#a^2 - #b^2', '(#a + #b)(#a - #b)')
+
+
+export const hello = (num, root = 2, precision = 12) => {
+  // e.g 2^-3 = 1/(2^3)
+  const inv = root < 0
+  if (inv) {
+    root = -root
+  }
+
+  if (root === 0) {
+    throw new Error('Root must be non-zero')
+  }
+  if (num < 0 && (Math.abs(root) % 2 !== 1)) {
+    throw new Error('Root must be odd when a is negative.')
+  }
+
+  // Edge cases zero and infinity.
+  // e.g 0^3 = 0, 0^-3 = Infinity
+  if (num === 0) {
+    return inv ? Infinity : 0
+  }
+
+  if (num === 1) {
+    return 1
+  }
+
+  if (!isFinite(num)) {
+    return inv ? 0 : num
+  }
+
+  // Source: https://rosettacode.org/wiki/Nth_root#JavaScript
+  const n = root
+  const prec = precision
+
+  let x = 1 // Initial guess
+  for (let i = 0 ; i < prec ; i++) {
+    x = 1 / n * ((n - 1) * x + (num / Math.pow(x, n - 1)))
+  }
+
+  return inv ? 1 / x : x
+}
+
+// TODO: handle fractional coefficients
+// TODO: isPolynomialTerm should handle fractional coeffs
+
+// e.g. 4x^2 + 12x + 9 -> (2x + 3)^2
+export const FACTOR_PERFECT_SQUARE_TRINOMIALS = defineRule(
+  (node) => {
+    const isFactorable = canApplyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    if (isFactorable) {
+      const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+      return print(result.args[0]) === print(result.args[1])
+        ? {node} : null
+    }
+  },
+
+  (node) => {
+    const result = applyRule(FACTOR_SUM_PRODUCT_RULE, node)
+    return build.pow(result.args[0], build.number(2))
+  }
+)
+
+const getExponent = (node) => {
+    if (query.isNumber(node)) {
+        return 0
+    } else if (query.isIdentifier(node)){
+        return 1
+    } else if (query.isPow(node)) {
+        return query.getValue(node.args[1])
+    } else if (query.isMul(node)){
+        return getExponent(node.args[1])
+    } else if (query.isNeg(node)) {
+        const variable = node.args[0]
+        return getExponent(variable.args[1])
+    } else {
+        return null
+    }
+}
+
+// e.g. 2 + 3x^2 + 3x - 4x^3 -> -4x^3 + 3x^2 + 3x + 2
+export const REARRANGE_TERMS = defineRule(
+    (node) => {
+        if (query.isAdd(node)) {
+            return node.args.some(arg => {
+                return isPolynomialTerm(arg)
+            }) ? {node} : null
+        }
+    },
+
+    (node) => {
+        const ordered = node.args.sort(function(a,b) {return getExponent(b) - getExponent(a)})
+        return build.add(...ordered)
+    }
+)
+
+const getFactors = (num) => {
```

Comment:
This should maybe be called `getFactorPairs` or add a comment describing what it does.  I was confused b/c I thought calling `getFactors(12)` would return `[1, 2, 3, 4, 6, 12]` but in fact it returns `[[1, 12], [2, 6], [3, 4]]`.